### ISSUE TITLE
[question] Is example for `Implementing the scm feature` (migration to Conan 2.0) useful?

### ISSUE THREAD

**Issue by gpxricky** [conan-io/conan#10777]
Hi Conans,

I would like to discuss the example and approach that is given on https://docs.conan.io/en/latest/reference/conanfile/tools/scm/git.html#example-implementing-the-scm-feature:

```
import os
from conan import ConanFile
from conan.tools.scm import Git
from conan.tools.files import load, update_conandata


class Pkg(ConanFile):
    name = "pkg"
    version = "0.1"

    def export(self):
        git = Git(self, self.recipe_folder)
        scm_url, scm_commit = git.get_url_and_commit()
        # we store the current url and commit in conandata.yml
        update_conandata(self, {"sources": {"commit": scm_commit, "url": scm_url}})

    def layout(self):
        self.folders.source = "."

    def source(self):
        # we recover the saved url and commit from conandata.yml and use them to get sources
        git = Git(self)
        sources = self.conan_data["sources"]
        git.clone(url=sources["url"], target=".")
        git.checkout(commit=sources["commit"])
```

I tried to implement the proposal for a Conan 1.x recipe, but it has some limitations I would like you to know:

- Git repositories may require authentication which is a little bit tricky with the code and Git class mentioned above (you need to add it dynamically to the url)
- Git raises an error that the source folder is not empty because some export files (e.g. the `conandata.yml`) are magically added

Furthermore the new approach clones the repository again when using `conan create` which may take much time (we have big repositories here) and prevents me from running this step offline (although package development flow should be fine). Is this really desired here?

My personal feeling is that the `scm` attribute is working fine and much better at present than the Conan 2.0 approach. In general I like the new idea and would ideally add it already to my Conan 1.X recipes, but the drawbacks and limitations mentioned above are a blocker for me.

What do you think?

Best regards,
Markus

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @gpxricky 

> Git repositories may require authentication which is a little bit tricky with the code and Git class mentioned above (you need to add it dynamically to the url)

We have intentionally removed the user/password management in URLs in this first iteration. It is something that more and more Git server providers are running away, it seems it is not the best practice nowadays. If you still need to do it, please report details on how/why, and we will consider adding support. If you can file a separate, dedicated issue with these details, that would help.

> Git raises an error that the source folder is not empty because some export files (e.g. the conandata.yml) are magically added

This is curious. The example in the docs is copy&pasted from a test that is running on the test suite, and it seems to be working fine. Are you sure you are running the code as documented, no other change? In that case, that needs to be investigated.

> Furthermore the new approach clones the repository again when using conan create which may take much time (we have big repositories here) and prevents me from running this step offline (although package development flow should be fine). Is this really desired here?

Can you please give some order of magnitude? Because even for a large repo, it would typically be only a small fraction of the build time, which would be much longer.

This is also done intentionally, the local copy was also quite dirty and fragile. It even had some very nasty bugs that it was re-creating a package from a completely obsolete user folder, very unexpected (and almost impossible to solve with the previous approach). The idea is that with the new local flow, the ``conan create`` should happen way less often, and only for very last mile testing. There are still improvements that could be done to get a faster ``Git.clone()`` in the ``source()`` method, which I guess is your performance concerns goes? Applying a ``--depth`` or something could help. In any case, the idea is that if you are running a ``conan create`` in your machine, it should do the same process that is doing in the CI machine. And the CI machine will need access to the git repo to clone the sources anyway. There might be easy ways to change that behavior, for example, using a ``conf`` we could always pick the local folder URL instead of the ``origin`` remote, but then there will be no guarantee that that works when it gets to the CI machine.

Take into account that this is the initial iteration, and of course there is space for improvements. So far we are prioritizing correctness/robustness. In short:
- Authentication support: can be added, please report details.
- Git clone raising because file, could be a bug and needs to be checked. Please check above comment on tests. 
- Git.clone() speed might be increased adding ``--depth``

Despite you feel it was better in ``scm`` in 1.X, it was simply because it was there for more time, but:
- There were some security flaws with credentials management in the past.
- There were serious bugs with the local copy of files, that were almost impossible to get rid of
- It was less flexible. The new approach has proven to be able to handle cases like the mono-repo that were impossible with the ``scm``.

**Comment by dornbirndevelops** [comment]
The only thing I don't necessarily like about the new approach is that I cannot `export` a dirty state anymore (not even with `--ignore-dirty`). This is a bit cumbersome when developing a component. The `scm` feature just gave me a warning saying that I cannot upload with auto values. That being said, I think the export and source function seems like a good idea once I can wrap my head around a new way of working... (I think `editables` should solve this)

**Comment by gpxricky** [comment]
Hi @memsharded,

you are right. I was backporting the approach to Conan 1.45.0 (using the old Git tool) which still copies export files to the source folder. With Conan 1.46.0 the example works fine here. Sorry for that confusion.

The Git repositories in my environment are not public, but protected by authentication. I don't know why, but Git commands in Conan do not seem to use the credential manager - which would be an option to bypass the authentication request. Anyway, as we are running inside a Docker environment it might not even be the case that the credential manager is configured or credentials are set already.
That was also the reason why I've written a small wrapper for the old Git tool to better support remote operations:

```
from conans import tools

class Git(tools.Git):
    """Wrapper for Conan's old Git tool (version 1.45.0).
    """

    def run(self, command):
        # Provide the username and password as environment variables in
        # order to hide them in the Git command (e.g. in case an error
        # message is raised that contains the command line call)
        env_vars = {}
        credential_helper = ""

        if self._username is not None and self._password is not None:
            env_vars.update({
                "GIT_USERNAME": self._username,
                "GIT_PASSWORD": self._password
            })
            credential_helper = '-c "credential.helper=!p() { ' \
                'echo "username=$GIT_USERNAME"; ' \
                'echo "password=$GIT_PASSWORD"; ' \
            '}; p" '

        with tools.environment_append(env_vars):
            command = credential_helper + command
            return super().run(command)
```

Regarding `clone` performance: we have repositories here with GBs full of files - some are using LFS, but not all (anyway that doesn't really matter here).  Cloning in home office via VPN takes a lot of time while the local build is just about 2 minutes.

On purpose, we use the `scm` attribute instead of the `export_sources` attribute, but have the sources and recipe located in the same repository. Therefore the CI build is exactly the same as my local build - it clones the repository once and then calls `conan create` - which copies the files magically to the Conan cache. Or did I miss something here?

I fully agree that the new approach is more flexible and has its advantages. Don't get my wrong, I know it is the initial state and I like it. I would just like to discuss some ideas or optimizations.

I don't know how (when being exported there is no chance to get back to the local folder, isn't it?), but maybe it is possible to specify a `reference` for the clone operation that points to the local recipe folder or use the `conf` approach you have mentioned? That would also enable to use of local commits that are not pushed to the remote yet. But it might also be the case that our development environment is too special for adding such a rare case officially to Conan - which I would fully understand too.

**Comment by memsharded** [comment]
@dornbirndevelops 

> The only thing I don't necessarily like about the new approach is that I cannot export a dirty state anymore (not even with --ignore-dirty). This is a bit cumbersome when developing a component. 

We are also aware of this one. We decided for the current approach because:
- The ``editable`` packages are orders of magnitude better. For the most part, it shouldn't be necessary to be exporting dirty state to test things (consumed by other packages), because you will be able to test those changes locally, and much faster (incremental build). We even added support for ``conan .... --build=editable`` for downstream consumer of editables, so it is possible to do full ordered builds.
- It is still possible to export dirty local state. Doing a local commit will work, and will return the URL of the local folder. Local commits can be easily reverted after the testing, it is not necessary to push the changes to remotes.
- The flow is now 100% configurable in the methods. The ``Git.get_url_and_commit()`` is just a higher level helper, but there are helpers for getting the commit, the dirty state, the local folder and the remote URL... You can write whatever you need for your case in the ``export()`` method in just a couple of lines. For example, you can even do a full ``copy()`` of your local changes if you detect that the repo is dirty.

>The scm feature just gave me a warning saying that I cannot upload with auto values. That being said, I think the export and source function seems like a good idea once I can wrap my head around a new way of working... (I think editables should solve this)

We will add some warnings/errors in the ``Git.get_url_and_commit()`` when possible, to be a bit more informative, but things should be clearly warned at ``export`` stage, and not wait until the ``upload`` stage.

**Comment by memsharded** [comment]
@gpxricky 

> That was also the reason why I've written a small wrapper for the old Git tool to better support remote operations

How did you inject that wrapper? Because that was not obvious in the ``scm`` approach, were you monkey patching? The current approach is exactly to provide this use case better, you can more elegantly replace the built-in Git helper for your own one. 
Also, I wouldn't mind to add built-in functionality to the new ``Git`` helper, specifically, if the credentials can be injected via env-var **name** like --password=$GIT_PASSWORD, instead of env-var value, which is always problematic and unsafe, because the password can easily leak in logs (and this was something that the previous Git integration was doing, replacing the password in the URL directly, then trying to hide it in output)

> The Git repositories in my environment are not public, but protected by authentication. I don't know why, but Git commands in Conan do not seem to use the credential manager - which would be an option to bypass the authentication request.

This is one of the things that should be investigated. If you can report a bit more (OS, credential manager, git version, etc), we should investigate why Git commands are not using it, they certainly should.


> I don't know how (when being exported there is no chance to get back to the local folder, isn't it?), but maybe it is possible to specify a reference for the clone operation that points to the local recipe folder or use the conf approach you have mentioned?

Yes, the ``self.recipe_folder`` for example is available in the ``export()`` method, and as commented in the previous comment, it can be saved in the conandata.yml, for example if you detect the operation is dirty, or simply because you want, or controlled by a ``conf``. And it can still save the ``url`` and ``commit``, for full reproducibility (like when building as dependency with ``conan install --build``). Then the ``source()`` method can use that saved folder in the conandata to do a ``copy()`` instead of a clone, keeping the same efficiency, and avoiding the extra clone. 

In short, it is still possible to implement the previous approach avoiding the extra clone, just more explicitly (which is a big gain for understanding what is happening, debugging and maintaining).

We will probably be adding more examples and some features to the docs as the feature matures.

**Comment by gpxricky** [comment]
@memsharded 

Do you mean something like this?

```
from conan.tools import scm
import os

class Git(scm.Git):
    def __init__(self, conanfile, folder=".",
            username_env_var=None,
            password_env_var=None):
        self._username_env_var = username_env_var
        self._password_env_var = password_env_var
        super().__init__(conanfile, folder)

    def _run(self, cmd):
        credential_helper = self._get_credential_helper()
        if credential_helper is not None:
            return super()._run(f"{credential_helper} {cmd}")
        return super()._run(cmd)

    def _get_credential_helper(self):
        user_var = self._username_env_var
        password_var = self._password_env_var

        if user_var is not None and password_var is not None:
            if user_var in os.environ and password_var in os.environ:
                return \
                    f'-c "credential.helper=!p() {{ ' \
                       f'echo "username=${user_var}"; ' \
                        f'echo "password=${password_var}"; ' \
                    f'}}; p"'

        return None
```

**Comment by memsharded** [comment]
@gpxricky Yes, I think we can make it built-in part of the ``Git`` helper, if we get enough understanding of the mechanism, and if it is low risk. So a couple of things:

- Is there some docs link? I cannot find docs for the ``!p() {}`` syntax. 
- Some general assesment on security risks? 

Some more feedback from other users regarding other potential features regarding credentials would help to implement the right abstraction for it.

**Comment by gpxricky** [comment]
@memsharded The `!p() {}` is just a special syntax that is supported by Git to include shell snippets (check rule no. 1 in https://git-scm.com/docs/gitcredentials#_custom_helpers). Honestly just the `!` is relevant here, the remaining part will be interpreted as shell snippet.

Anyway, the solution didn't work stable for me, so I ended up using the following solution (works fine for Windows and Linux):

```
class GitCredentialHelper(object):
    def __init__(self, username_env_var=None,
            password_env_var=None):
        self._username_env_var = username_env_var
        self._password_env_var = password_env_var
        self._create_temp_file()

    @property
    def use_credential_helper(self):
        user_var = self._username_env_var
        password_var = self._password_env_var

        if user_var is not None and password_var is not None:
            if user_var in os.environ and password_var in os.environ:
                return True

        return False

    @property
    def cli_arguments(self):
        if self._temp_file is not None:
            # Convert backslashes in the path to forward slashes on
            # Windows OS (double backslashes would also work fine)
            temp_file = self._temp_file.replace("\\", "/")
            # Whatever the exact reason is, but any configured
            # credential helper needs to be disabled first (otherwise
            # Git will first ask the configured credential helper for
            # the username and password), before the new one is set
            return f'-c credential.helper="" ' \
                f'-c credential.helper="{temp_file}"'

        return ""

    def close(self):
        if self._temp_file is not None:
            Path(self._temp_file).unlink()
            self._temp_file = None

    def _create_temp_file(self):
        self._temp_file = None

        if self.use_credential_helper:
            # We handle the deletion of the temporary file at our
            # own in order to prevent "Text file busy" errors on Windows
            with tempfile.NamedTemporaryFile(mode="w", delete=False) as temp:
                self._temp_file = temp.name
                temp.write(self._get_credential_helper_content())

            Path(self._temp_file).chmod(0o777)

    def _get_credential_helper_content(self):
        content = '\n'.join([
            f'#!/bin/bash',
            f'if [ "$1" == "get" ]; then',
            f'  echo "username=${self._username_env_var}"',
            f'  echo "password=${self._password_env_var}"',
            f'fi'
        ]) + '\n'

        return content


class Git(scm.Git):
    """Wrapper for Conan's Git tool that supports to pass an environment
    variable that stores the username and password.
    """
    def __init__(self, conanfile, folder=".",
            username_env_var=None,
            password_env_var=None):

        self._username_env_var = username_env_var
        self._password_env_var = password_env_var
        super().__init__(conanfile, folder)

    def _run(self, cmd):
        with self._credential_helper() as creds:
            if creds.use_credential_helper:
                return super()._run(f"{creds.cli_arguments} {cmd}")
            return super()._run(cmd)

    @contextmanager
    def _credential_helper(self):
        credential_helper = GitCredentialHelper(
            self._username_env_var,
            self._password_env_var
        )

        try:
            yield credential_helper
        finally:
            credential_helper.close()
```

**Comment by puetzk** [comment]
> Git raises an error that the source folder is not empty because some export files (e.g. the conandata.yml) are magically added


>This is curious. The example in the docs is copy&pasted from a test that is running on the test suite, and it seems to be working fine. 

This is fallout of #10625 - in conan 1.x, the example as given does not work, because conan dumps the exports() files (conanfile.py, conandata.yml, and anything else export copied to the recipe_folder) into the source dir, before it calls source(). If the recipe **also** overrides `def layout()`, then it no longer does this (and 2.x doesn't do so at all).

**Comment by puetzk** [comment]
> Some more feedback from other users regarding other potential features regarding credentials would help to implement the right abstraction for it.

The abstraction I'd like to see for it would be for conan.tools.scm.Git to sneak an addition of `--reference repository` in as part of `conan create`. Either on the `git clone`, appending to `.git/objects/info/alternates` in between `init` and `fetch`, or even just `GIT_ALTERNATE_OBJECT_DIRECTORIES` in the environment when it calls git.exe

That preserves the purity aspect (and would not allow a *dirty* build), but it doesn't actually have to re-download anything, nor duplicate any of the packfiles files locally.

With Git.clone() it would still need to talk to the remote a little bit (so it would need credentials), just to obtain the complete list of branches/tags/other named refs.

But I think Git.fetch_commit would be be able to stay completely offline in this scenario, since checking out a hash already available via .git/objects/info/alternates should just work (without needing any call to `git fetch`). It doesn't seem that `git fetch <origin> <sha>` make itself a no-op, it still checks whether the remote actually has that hahs. So to work offline,  fetch_commit might need to do a `git rev-parse commit^{tree}` to check whether .git already has this commit's tree (and then if not, use fetch to get it).

That leaves the questions of how `conan create` would collude to smuggle a local .git/object folder between `get_url_and_commit()` and `clone()` or `fetch_commit()`. I see a couple plausible ideas

1. serialize the path into the local cache,  in a way that upload() does not preserve, like conan 1.x did with its scm_source folder
2. ephemeral in-memory cooperation between conan.tools.scm.Git(ConanFile) and the ConanFile class - any repo you did get_url_and_commit() on gets becomes a place that clone() might use with `--reference <x> --detach`, and `fetch_commit` could just reference directly.
4. A mechanism like core.download:download_cache, where conan keeps a repo consisting of a long list of .git/objects/info/alternates pointing to places conan.tools.scm.Git.get_commit returned something from). This would seem to rapidly accumulate broken links, though...

IMO the use case to favor here is `conan create`, so I'd lean toward envvar magic where repos referenced during export() end up listed in `GIT_ALTERNATE_OBJECT_DIRECTORIES` during source(), fetch_commit learns to check when it can skip `git fetch`, and nothing gets recorded anywhere else. That's a pretty pure play to just speeds things up / work offline, with no other semantic impacts.

**Comment by memsharded** [comment]
Hi all @puetzk , @gpxricky , @dornbirndevelops 

Some time has passed since the original posting, and I think there are many different topics in this thread, which makes it a bit cumbersome for us to prioritize and take actions. 

One of the original issues was related to credentials. We introduced a ``source_credentials.json`` file for ``download()`` and ``get()`` helpers, but only for downloads, it is not used for git clones. We have been trying to recommend to use other orthogonal built-in auth mechanisms for Git like using ssh keys, and we have been reported it working good, and also we have been reported about succesful usage of GIT_ASKPASS. In any case, we still believe that builtin mechanisms that let Conan be agnostic of the credentials for Git are better. Developers would already have them in their machines to do their jobs, and CI systems already need some kind of built-in Git credentials even before Conan starts, because cloning sources is one of the primary and first things a CI needs to do anyway, using or not Conan.

Then, it seems there are also other comments related to performance. In that regard the latest comments of @puetzk seems going in this direction, but not fully clear to me. For performance issues, we also added the ``Git.fetch_commit()`` helper that would fetch just 1 commit, we thought that it could provide some optimization. We would like to know the order of magnitude of approaches using this ``fetch_commit()`` before thinking about further optimizations. 

Maybe it would be better to split more focused tickets for each one of these discussions, I'd recommend that.

**Comment by memsharded** [comment]
Hi all,

I am closing this ticket as staled/inactive for a long time, but the last comment is still valid, this ticket had different topics, some of them might have been solved. So please open new dedicated tickets for any further question, issue or feedback. Many thanks!
